website/blog/2024/11-10-web-scraping-tips/index.md

```diff
@@ -0,0 +1,293 @@
+
+After identifying the endpoints you need to extract the target data, make sure you get a correct response when making a request. If you get a response from the server other than 200, or data different from expected, then you need to figure out why. Here are some possible reasons:
+
+- You need to pass some parameters, for example cookies, or specific technical headers
+- The site requires that when accessing this endpoint, there is a corresponding `Referrer` header
+- The site expects that the headers will follow a certain order. I've encountered this only a couple of times, but I have encountered it
+- The site uses protection against web scraping, for example with `TLS fingerprint`
+
+And many other possible reasons, each of which requires separate analysis.
+
+## 10. Experiment with request parameters
+
+Explore what results you get when changing request parameters, if any. Some parameters may be missing but supported on the server side. For example, `order`, `sort`, `per_page`, `limit`, and others. Try adding them and see if the behavior changes.
+
+This is especially relevant for sites using [`graphql`](https://graphql.org/)
+
+Let's consider this [`example`](https://restoran.ua/en/posts?subsection=0)
+
+If you analyze the site, you'll see a request that can be reproduced with the following code, I've formatted it a bit to improve readability:
+
+```python
+import requests
+
+url = "<https://restoran.ua/graphql>"
+
+data = {
+    "operationName": "Posts_PostsForView",
+    "variables": {"sort": {"sortBy": ["startAt_DESC"]}},
+    "query": """query Posts_PostsForView(
+    $where: PostForViewWhereInput,
+    $sort: PostForViewSortInput,
+    $pagination: PaginationInput,
+    $search: String,
+    $token: String,
+    $coordinates_slice: SliceInput,
+    {
+        PostsForView(
+                where: $where
+                sort: $sort
+                pagination: $pagination
+                search: $search
+                token: $token
+                ) {
+                        id
+                        title: ukTitle
+                        summary: ukSummary
+                        slug
+                        startAt
+                        endAt
+                        newsFeed
+                        events
+                        journal
+                        toProfessionals
+                        photoHeader {
+                            address: mobile
+                            __typename
+                            }
+                        coordinates(slice: $coordinates_slice) {
+                            lng
+                            lat
+                            __typename
+                            }
+                        __typename
+                    }
+    }"""
+}
+
+response = requests.post(url, json=data)
+
+print(response.json())
+```
+
+Now I'll update it to get results in 2 languages at once, and most importantly, along with the internal text of the publications:
+
+```python
```

Comment:
I'd rather see what was added than see the whole script again... consider the `diff` Markdown syntax highlighter or Docusaurus' [Line Highlighting](https://docusaurus.io/docs/markdown-features/code-blocks#line-highlighting) (I'd actually prefer this).